client/allocrunner/taskrunner/identity_hook.go

```diff
@@ -65,52 +70,138 @@ func (*identityHook) Name() string {
 	return "identity"
 }
 
-func (h *identityHook) Prestart(context.Context, *interfaces.TaskPrestartRequest, *interfaces.TaskPrestartResponse) error {
+func (h *identityHook) Prestart(ctx context.Context, _ *interfaces.TaskPrestartRequest, _ *interfaces.TaskPrestartResponse) error {
 
 	// Handle default workload identity
 	if err := h.setDefaultToken(); err != nil {
 		return err
 	}
 
+	// Track first run signals from watchers
+	firstRunCh := make(chan struct{}, len(h.task.Identities))
+
 	// Start token watcher loops
 	for _, widspec := range h.task.Identities {
 		w := widspec
-		go h.watchIdentity(w)
+		go h.watchIdentity(w, firstRunCh)
+	}
+
+	// Don't block indefinitely for identities
+	deadlineTimer := time.NewTimer(time.Minute)
+	defer deadlineTimer.Stop()
+
+	// Wait until every watcher ticks the first run chan
+	for i := range h.task.Identities {
+		select {
+		case <-firstRunCh:
+			// Identity fetched, loop
+		case <-deadlineTimer.C:
```

Comment:
kind of makes me think there should be a context-aware version of WaitGroup

Reply:
100% what I needed